velox/exec/benchmarks/OrderByBenchmark.cpp

```diff
@@ -0,0 +1,140 @@
+#include "velox/exec/tests/utils/PlanBuilder.h"
+
+using namespace facebook::velox;
+using namespace facebook::velox::exec;
+
+namespace {
+
+struct TestCase {
+  explicit TestCase(
+      const std::string& testName,
+      size_t numRows,
+      const RowTypePtr& rowType,
+      int numKeys)
+      : testName{testName},
+        numRows{numRows},
+        rowType{rowType},
+        numKeys{numKeys} {}
+
+  std::string testName;
+  size_t numRows;
+  RowTypePtr rowType;
+  int numKeys;
+};
+
+class OrderByBenchmark {
+ public:
+  void addBenchmark(
+      const std::string& testName,
+      size_t numRows,
+      const RowTypePtr& rowType,
+      int32_t iterations,
+      int numKeys) {
+    auto testCase =
+        std::make_unique<TestCase>(testName, numRows, rowType, numKeys);
+    {
+      folly::addBenchmark(
+          __FILE__,
+          "OrderBy_" + testCase->testName,
+          [test = testCase.get(),
+           iterations = std::max(1, iterations / 10),
+           this]() {
+            auto plan = makeOrderByPlan(*test);
+            uint64_t inputNanos = 0;
+            uint64_t outputNanos = 0;
+            uint64_t finishNanos = 0;
+            auto start = getCurrentTimeMicro();
+            for (auto i = 0; i < iterations; ++i) {
+              std::shared_ptr<Task> task;
+              test::AssertQueryBuilder(plan).copyResults(pool_.get(), task);
+              auto stats = task->taskStats();
+              for (auto& pipeline : stats.pipelineStats) {
+                for (auto& op : pipeline.operatorStats) {
+                  if (op.operatorType != "OrderBy") {
+                    continue;
+                  }
+                  inputNanos += op.addInputTiming.wallNanos;
+                  finishNanos += op.finishTiming.wallNanos;
+                  outputNanos += op.getOutputTiming.wallNanos;
+                }
+              }
+            }
+            uint64_t total = getCurrentTimeMicro() - start;
+            std::cout << "Total " << succinctMicros(total) << " Input "
+                      << succinctNanos(inputNanos) << " Output "
+                      << succinctNanos(outputNanos) << " Finish "
+                      << succinctNanos(finishNanos) << std::endl;
+            return 1;
+          });
+    }
+    testCases_.push_back(std::move(testCase));
+  }
+
+ private:
+  core::PlanNodePtr makeOrderByPlan(TestCase& test) {
+    folly::BenchmarkSuspender suspender;
+    std::vector<RowVectorPtr> vectors;
+    vectors.emplace_back(OrderByBenchmarkUtil::fuzzRows(
+        test.rowType, test.numRows, test.numKeys, pool_.get()));
+
+    std::vector<std::string> keys;
+    keys.reserve(test.numKeys);
+    for (auto i = 0; i < test.numKeys; i++) {
+      keys.emplace_back(fmt::format("c{} ASC NULLS LAST", i));
+    }
+
+    auto planNodeIdGenerator = std::make_shared<core::PlanNodeIdGenerator>();
+    core::PlanNodeId orderNodeId;
+    return test::PlanBuilder(planNodeIdGenerator)
```

Comment:
This plan runs orderBy and then copies results. The cost of copy is probably significant, hence, I suggest to add a node that discards the results, i.e. count(1) aggregation or limit(1) or similar. Alternatively, copyResults can be replaced with a new method 'countResults' that simply counts number of rows produced by the query. 

Furthermore, it might be better to use Task::next API  to run query single-threaded. This allows to measure the cost of OrderBy more accurately.

Reply:
Thanks for catch this. Changed.